lib/correios/cep/web_service.rb

```diff
@@ -48,20 +44,31 @@ def build_request(zipcode)
         request = Net::HTTP::Post.new(uri.path)
         request['Content-Type'] = 'text/xml; charset=utf-8'
         request.body = request_body(zipcode)
+        Correios::CEP.log_request(request, uri.to_s)
+
         request
       end
 
       def request_body(zipcode)
-        '<?xml version="1.0" encoding="UTF-8"?>' +
-        '<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:cli="http://cliente.bean.master.sigep.bsb.correios.com.br/">' +
-           '<soapenv:Header />' +
-           '<soapenv:Body>' +
-              '<cli:consultaCEP>' +
-                "<cep>#{zipcode}</cep>" +
-              '</cli:consultaCEP>' +
-           '</soapenv:Body>' +
+        '<?xml version="1.0" encoding="UTF-8"?>' \
+        '<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:cli="http://cliente.bean.master.sigep.bsb.correios.com.br/">' \
+           '<soapenv:Header />' \
+           '<soapenv:Body>' \
+              '<cli:consultaCEP>' \
+                "<cep>#{zipcode}</cep>" \
+              '</cli:consultaCEP>' \
+           '</soapenv:Body>' \
         '</soapenv:Envelope>'
       end
+
+      def request_with_retry
+        retries ||= 0
+        response = yield
+        Correios::CEP.log_response(response)
+        response
+      rescue EOFError
+        retry if (retries += 1) < Correios::CEP.max_retries || raise
+      end
```

Comment:
This method really don't make a request. It just runs the block and retries in case of `EOFError` is raised. 
Who decides if the block is a request (could be anything) is the client code. 

Could you think in a better method name? 😉